clippy_lints/src/identity_conversion.rs

```diff
@@ -72,7 +72,12 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for IdentityConversion {
                     let a = cx.tables.expr_ty(e);
                     let b = cx.tables.expr_ty(&args[0]);
                     if same_tys(cx, a, b) {
-                        let sugg = snippet(cx, args[0].span, "<expr>").into_owned();
+                        let sugg = if in_macro(args[0].span) {
+                            snippet(cx, args[0].span.source_callsite(), "<expr>").into_owned()
```

Comment:
Can't we unconditionally call the source callsite methods? if so, can we move it into the snippet function so all lints benefit from it?

Reply:
Yea, I was thinking that this is probably becoming a recurring pattern. I will give it a try!

Reply:
One problem occurs when `span.source_callsite()` is called on a `format!` argument. This happens in `useless_format` lint (and a few other lints):

https://github.com/rust-lang-nursery/rust-clippy/blob/457e7f12e9fc028eae182f23b279194e5344b676/clippy_lints/src/format.rs#L78-L79

Unconditionally calling `source_callsite` will result in the wrong suggestion because the `source_callsite` exists and is the `format!` macro in this case:

```
help: consider using .to_string(): `format!("foo");.to_string()`
```

I think it still makes sense to have a separate `snippet_with_macro_callsite` function for the other cases, though.

Reply:
Makes sense. This generally only works if the entire situation is outside a macro, but most of our lints are skipping those situations via `in_macro` on the full span anyway.